[SPARK-14610][ML] Remove superfluous split for continuous features in decision tree training

sethah opened the pull request:
## What changes were proposed in this pull request?

A nonsensical split is produced from method `findSplitsForContinuousFeature` for decision trees. This PR removes the superfluous split and updates unit tests accordingly. Additionally, an assertion to check that the number of found splits is `> 0` is removed, and instead features with zero possible splits are ignored. 
## How was this patch tested?

A unit test was added to check that finding splits for a constant feature produces an empty array.

sethah committed:
remove extra split for continuous features

sethah committed:
cleanup

SparkQA commented:
**[Test build #55765 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55765/consoleFull)** for PR 12374 at commit [`98c31e9`](https://github.com/apache/spark/commit/98c31e9ca53de248c582cfd21959568912d8939e).

sethah commented:
cc @jkbradley 

This is a small PR that generally makes things _more correct_. But, I realize that this did not really have any adverse effects before, so I'll understand if this does not get merged. It is likely to be more of a problem when working on micro datasets. Although, there is a small change I included which I do think is incorrect and should be fixed, regarding how `possibleSplits` is defined in `findSplitsForContinuousFeatures`.

sethah commented on mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala:
```diff
@@ -966,9 +972,9 @@ private[spark] object RandomForest extends Logging {
 
       // if possible splits is not enough or just enough, just return all possible splits
-      val possibleSplits = valueCounts.length
+      val possibleSplits = valueCounts.length - 1
```
The number of possible _bins_ should be `valueCounts.length`, and the number of possible _splits_ should therefore be `valueCounts.length - 1`.

SparkQA commented:
**[Test build #55765 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55765/consoleFull)** for PR 12374 at commit [`98c31e9`](https://github.com/apache/spark/commit/98c31e9ca53de248c582cfd21959568912d8939e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55765/
Test FAILed.

sethah committed:
unit test failure

SparkQA commented:
**[Test build #55768 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55768/consoleFull)** for PR 12374 at commit [`0a26a1f`](https://github.com/apache/spark/commit/0a26a1f84af3ba77cd798bcb27661a64170a7709).

SparkQA commented:
**[Test build #55768 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55768/consoleFull)** for PR 12374 at commit [`0a26a1f`](https://github.com/apache/spark/commit/0a26a1f84af3ba77cd798bcb27661a64170a7709).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55768/
Test PASSed.

SparkQA commented:
**[Test build #59138 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59138/consoleFull)** for PR 12374 at commit [`0a26a1f`](https://github.com/apache/spark/commit/0a26a1f84af3ba77cd798bcb27661a64170a7709).

SparkQA commented:
**[Test build #59138 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59138/consoleFull)** for PR 12374 at commit [`0a26a1f`](https://github.com/apache/spark/commit/0a26a1f84af3ba77cd798bcb27661a64170a7709).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59138/
Test PASSed.

sethah committed:
handle empty case

sethah commented:
cc @MechCoder @MLnick Could you take a look?

sethah commented on mllib/src/test/scala/org/apache/spark/ml/tree/impl/RandomForestSuite.scala:
```diff
@@ -137,14 +137,47 @@ class RandomForestSuite extends SparkFunSuite with MLlibTestSparkContext {
+      assert(splits === Array[Double]())
+      val splitsEmpty =
+        RandomForest.findSplitsForContinuousFeature(featureSamplesEmpty, fakeMetadata, 0)
+      assert(splitsEmpty === Array[Double]())
```
This test would have failed before due to the assertion that `splits.length > 0`.

sethah commented on mllib/src/test/scala/org/apache/spark/ml/tree/impl/RandomForestSuite.scala:
```diff
@@ -137,14 +137,47 @@ class RandomForestSuite extends SparkFunSuite with MLlibTestSparkContext {
+    }
+  }
+
+  test("train with constant features") {
```
This test would have failed before due to the assertion that `splits.length > 0`.

MechCoder replied:
"train with constant features" -> "train with constant continuous features"?

sethah replied:
This test is not specific to continuous features.

sethah commented on mllib/src/test/scala/org/apache/spark/ml/tree/impl/RandomForestSuite.scala:
```diff
@@ -137,14 +137,47 @@ class RandomForestSuite extends SparkFunSuite with MLlibTestSparkContext {
+      val featureSamplesEmpty = Array.empty[Double]
+      val splits = RandomForest.findSplitsForContinuousFeature(featureSamples, fakeMetadata, 0)
+      assert(splits === Array[Double]())
+      val splitsEmpty =
```
Although it is not currently possible for the `findSplitsForContinuousFeature` method to receive an empty array, we still handle it. I appreciate feedback on this.

MechCoder replied:
When will this ever happen, or in other words what corner case does this fix?

sethah replied:
The iterable passed to `findSplitsForContinuousFeature` is the result of a `groupByKey`, so I don't see how this could ever happen. Still, if this method is used elsewhere ever in the future, it would fail with `java.lang.UnsupportedOperationException`. I'm open to changing this.

SparkQA commented:
**[Test build #60975 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60975/consoleFull)** for PR 12374 at commit [`a42c126`](https://github.com/apache/spark/commit/a42c126774db1c0dc8ec5b1d5019e99bf37be402).

SparkQA commented:
**[Test build #60975 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60975/consoleFull)** for PR 12374 at commit [`a42c126`](https://github.com/apache/spark/commit/a42c126774db1c0dc8ec5b1d5019e99bf37be402).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60975/
Test FAILed.

sethah committed:
add instrumentation variable to test

SparkQA commented:
**[Test build #60979 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60979/consoleFull)** for PR 12374 at commit [`1b7f826`](https://github.com/apache/spark/commit/1b7f826c3ee6512affbac6edb4e6ac32d1dfa9f2).

SparkQA commented:
**[Test build #60979 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60979/consoleFull)** for PR 12374 at commit [`1b7f826`](https://github.com/apache/spark/commit/1b7f826c3ee6512affbac6edb4e6ac32d1dfa9f2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60979/
Test PASSed.

MechCoder commented on mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala:
```diff
@@ -712,17 +712,23 @@ private[spark] object RandomForest extends Logging {
-              (splitIdx, gainAndImpurityStats)
-            }.maxBy(_._2.gain)
-          (splits(featureIndex)(bestFeatureSplitIndex), bestFeatureGainStats)
+          if (numSplits == 0) {
```
This seems slightly hacky to me. What is your opinion about doing filtering out the feature indices that have zero splits (something similar to this)?

```scala
val validFeaturesSplits = Range(0, binAggregates.metadata.numFeaturesPerNode).filter { featureIndexIdx =>
  val featureIndex = if (featuresForNode.nonEmpty) {
    featuresForNode.get.apply(featureIndexIdx)
  } else {
    featureIndexIdx
  }
  binAggregates.metadata.numSplits(featureIndex) != 0
}
```

That will prevent code-rewrite for this corner-case in PR's such as https://github.com/apache/spark/pull/13959 and https://github.com/apache/spark/pull/8540

sethah replied:
I agree. I modified your suggestion to work with a view, so we don't allocate unnecessary memory.

MechCoder commented on mllib/src/test/scala/org/apache/spark/ml/tree/impl/RandomForestSuite.scala:
```diff
@@ -114,7 +114,7 @@ class RandomForestSuite extends SparkFunSuite with MLlibTestSparkContext {
       val featureSamples = Array(1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 3).map(_.toDouble)
       val splits = RandomForest.findSplitsForContinuousFeature(featureSamples, fakeMetadata, 0)
-      assert(splits.length === 3)
+      assert(splits.length === 2)
```
I would check the splits explicitly, that is the `Array(1, 2)`

sethah replied:
Done.

MechCoder commented on mllib/src/test/scala/org/apache/spark/ml/tree/impl/RandomForestSuite.scala:
```diff
@@ -137,14 +137,47 @@ class RandomForestSuite extends SparkFunSuite with MLlibTestSparkContext {
+          maxDepth = 2,
+          numClasses = 100,
+          maxBins = 100,
+          categoricalFeaturesInfo = Map(0 -> 2, 1 -> 5))
```
I would just remove `categoricalFeaturesInfo`

sethah replied:
I am using this to not only test continuous features, but categorical ones as well. I think we should leave it.

MechCoder commented on mllib/src/test/scala/org/apache/spark/ml/tree/impl/RandomForestSuite.scala:
```diff
@@ -137,14 +137,47 @@ class RandomForestSuite extends SparkFunSuite with MLlibTestSparkContext {
+          OldAlgo.Classification,
+          Gini,
+          maxDepth = 2,
+          numClasses = 100,
```
`numClasses=100`?

sethah replied:
Not sure what this comment means, but if it's a reference to style, then the current syntax is correct. Unlike Python, named parameters are provided with spaces. See [here](http://docs.scala-lang.org/sips/completed/named-and-default-arguments.html)

MechCoder replied:
My concern was that `numClasses=100` is set here, but the way they data is initialized suggests that `numClasses=1`. If we ever write code that validates the `numClasses` from data, these tests will break. (Similar concerns about `categoricalFeaturesInfo`).

MechCoder replied:
Is it the case that `numClasses` can be greater than the number of unique labels in the data. If yes, then ignore the comment.

sethah replied:
Good catch, the strategy was simply adapted from another test. I have changed it to 2, since the labels are zero indexed.

MechCoder commented:
@sethah Nice catch! This superfluous split seems to be only for continuous features in which the number of unique values - 1 is lesser than or equal to the number of splits. Can you update the PR title or description to reflect this change? Thanks!

MechCoder commented on mllib/src/test/scala/org/apache/spark/ml/tree/impl/RandomForestSuite.scala:
```diff
@@ -137,14 +137,47 @@ class RandomForestSuite extends SparkFunSuite with MLlibTestSparkContext {
+          maxBins = 100,
+          categoricalFeaturesInfo = Map(0 -> 2, 1 -> 5))
+    val Array(tree) = RandomForest.run(rdd, strategy, 1, "all", 42L, instr = None)
+    assert(tree.rootNode.impurity === -1.0)
```
Why is the impurity of the rootNode  "-1"? Since there is only class should it not be just zero?

sethah replied:
No, since the node found no valid split we flag the impurity as invalid. See [here](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/tree/Node.scala#L287)

MechCoder commented:
Outside of this PR, I would like to either:

1. Update the documentation of `findSplitsForContinuousFeature` to reflect that the return type is an array of thresholds, rather than an array of Splits.
2. Change the return type of `findSplitsForContinuousFeature` to return an array of splits directly.

(The 2nd one more preferable)

sethah committed:
address some review comments

SparkQA commented:
**[Test build #62106 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62106/consoleFull)** for PR 12374 at commit [`3bb28fe`](https://github.com/apache/spark/commit/3bb28fe671b7f2aeac486ea8eb87809a997e6245).

sethah commented:
@MechCoder I addressed your comments. I updated the scala doc for `findSplitsForContinuousFeature` to reflect the return type. I think it's fine to simply fix the doc for now. Let me know if you see anything else.

SparkQA commented:
**[Test build #62106 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62106/consoleFull)** for PR 12374 at commit [`3bb28fe`](https://github.com/apache/spark/commit/3bb28fe671b7f2aeac486ea8eb87809a997e6245).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62106/
Test PASSed.

sethah committed:
cleanup

SparkQA commented:
**[Test build #62562 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62562/consoleFull)** for PR 12374 at commit [`eddac63`](https://github.com/apache/spark/commit/eddac63af93c0ba8ff0ffb218029263136253c24).

SparkQA commented:
**[Test build #62562 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62562/consoleFull)** for PR 12374 at commit [`eddac63`](https://github.com/apache/spark/commit/eddac63af93c0ba8ff0ffb218029263136253c24).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62562/
Test PASSed.

MechCoder commented on mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala:
```diff
@@ -692,14 +692,20 @@ private[spark] object RandomForest extends Logging {
+    val validFeatureSplits =
+      Range(0, binAggregates.metadata.numFeaturesPerNode).view.map { featureIndexIdx =>
+        if (featuresForNode.nonEmpty) {
+          (featureIndexIdx, featuresForNode.get.apply(featureIndexIdx))
```
Is the `apply` here redundant?

sethah replied:
I don't think so. The alternative is `featuresForNode.get(featureIndexIdx)` which is misleading even though it does work. It looks like you are calling a function `get` and passing `featureIndexIdx` as an argument. Explicit apply seems clearer.

MechCoder commented:
LGTM

sethah committed:
update numClasses in test

SparkQA commented:
**[Test build #62830 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62830/consoleFull)** for PR 12374 at commit [`3c73726`](https://github.com/apache/spark/commit/3c73726ae986518bb9ef6fb4b0eb0fdb8e7e6616).

SparkQA commented:
**[Test build #62830 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62830/consoleFull)** for PR 12374 at commit [`3c73726`](https://github.com/apache/spark/commit/3c73726ae986518bb9ef6fb4b0eb0fdb8e7e6616).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62830/
Test PASSed.

SparkQA commented:
**[Test build #64417 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64417/consoleFull)** for PR 12374 at commit [`3c73726`](https://github.com/apache/spark/commit/3c73726ae986518bb9ef6fb4b0eb0fdb8e7e6616).

SparkQA commented:
**[Test build #64417 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64417/consoleFull)** for PR 12374 at commit [`3c73726`](https://github.com/apache/spark/commit/3c73726ae986518bb9ef6fb4b0eb0fdb8e7e6616).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64417/
Test PASSed.

sethah committed:
small cleanups

sethah commented:
ping @jkbradley or @yanboliang

SparkQA commented:
**[Test build #66676 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66676/consoleFull)** for PR 12374 at commit [`928a834`](https://github.com/apache/spark/commit/928a834a50d6921c825021fffc6f2a810122cb6c).

SparkQA commented:
**[Test build #66676 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66676/consoleFull)** for PR 12374 at commit [`928a834`](https://github.com/apache/spark/commit/928a834a50d6921c825021fffc6f2a810122cb6c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66676/
Test PASSed.

jkbradley commented:
LGTM
Merging with master
Thanks!

sethah commented:
Thanks @jkbradley!
